nixos/modules/services/web-apps/jitsi-meet.nix

```diff
@@ -0,0 +1,497 @@
+        enabled = true;
+        domain = cfg.hostName;
+        extraConfig = ''
+          authentication = "anonymous"
+          c2s_require_encryption = false
+          admins = { "focus@auth.${cfg.hostName}" }
+        '';
+      };
+      virtualHosts."auth.${cfg.hostName}" = {
+        enabled = true;
+        domain = "auth.${cfg.hostName}";
+        extraConfig = ''
+          authentication = "internal_plain"
+        '';
+      };
+    };
+    systemd.services.prosody.serviceConfig = mkIf cfg.prosody.enable {
+      EnvironmentFile = [ "${dataDir}/secrets" ];
+      SupplementaryGroups = [ "jitsi-meet" ];
+    };
+
+    users.groups.jitsi-meet = {};
+    systemd.tmpfiles.rules = [
+      "d '${dataDir}' 0750 root jitsi-meet - -"
+    ];
+
+    systemd.services.jitsi-meet-init-secrets = {
+      wantedBy = [ "multi-user.target" ];
+      before = [ "jicofo.service" "jitsi-videobridge2.service" ] ++ (optional cfg.prosody.enable "prosody.service");
+      serviceConfig = {
+        Type = "oneshot";
+      };
+
+      script = let
+        secrets = [ "VIDEOBRIDGE_SECRET" "JICOFO_COMPONENT_SECRET" "JICOFO_USER_SECRET" ];
+      in
+      ''
+        if [ ! -f ${dataDir}/secrets ]; then
+          ${concatMapStringsSep "\n" (var: ''
+            echo "${var}=$(tr -dc a-zA-Z0-9 </dev/urandom | head -c 64)" >> ${dataDir}/secrets
+          '') secrets}
+
+          chown root:jitsi-meet ${dataDir}/secrets
+          chmod 640 ${dataDir}/secrets
+        fi
+
+      ''
+      + optionalString cfg.prosody.enable ''
+        source ${dataDir}/secrets
+        ${config.services.prosody.package}/bin/prosodyctl register focus auth.${cfg.hostName} "$JICOFO_USER_SECRET"
+        ${config.services.prosody.package}/bin/prosodyctl register jvb auth.${cfg.hostName} "$VIDEOBRIDGE_SECRET"
+
+        cd ${config.services.prosody.dataDir}
+
+        # regenerate the certificates after 1 month - they are valid for a year, that leaves 11 months for service restart
+        find ${cfg.hostName}.crt -mtime +30 -delete || true
+
+        # generate self-signed certificates
+        if [ ! -f ${cfg.hostName}.crt ]; then
+          export PATH=$PATH:${pkgs.openssl.bin}/bin
+          rm -f {,auth.}${cfg.hostName}.{crt,key,cnf}
+          # not sure this one is even used but it avoids some errors in logs
+          echo | ${config.services.prosody.package}/bin/prosodyctl cert generate ${cfg.hostName}
+          # used for jicofo<->prosody user connection
+          echo | ${config.services.prosody.package}/bin/prosodyctl cert generate auth.${cfg.hostName}
+        fi
+
+        # don't know how to add it to the system-wide ceritificate bundle, instead convert it to java truststore and pass it to jicofo
+        rm -f ${dataDir}/prosody-ca.jks
+        ${pkgs.jre}/bin/keytool -importcert -file auth.${cfg.hostName}.crt -keystore ${dataDir}/prosody-ca.jks -storepass javapls -noprompt
+        chown root:jitsi-meet ${dataDir}/prosody-ca.jks
```

Comment:
Can we use the acme module to request proper cerificates here, instead of our own and having to fiddle with the java truststore?

Reply:
This is the difficult part IMHO.

In theory we can, though it requires the user to create DNS records which is currently not needed. For multi-host setup we'll probably end up with offering the choice of using acme or supplying custom certificates. Both require some manual work prior to deploying the module though.

Perhaps there could be third option to disable certificate verification, which should be fine for single-host setup. I didn't know how to do that when I wrote this truststore fiddling but now I do.

Reply:
For what it's worth, the current setup mimics exactly what both the official upstream package and the manual instructions do. I would certainly not use ACME as a default. There could be some value to being able to provide custom certificates for (for example) a system with _just_ a videobridge.

That would also require the ability to run the components separately, which should probably be in a separate PR. Without that separation, it doesn't make a lot of sense to move away from contained self-signed certs.